pkg/pillar/cmd/downloader/mdns.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+// github.com/grandcat/zeroconf: under MIT License
+
+package downloader
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/grandcat/zeroconf"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+// given a local server URL string, and returns the modified url with server IP address in it,
+// and bridge name and source IP address for connecting to the server.
+func findDSmDNS(dctx *downloaderContext, serverURL string) (string, string, net.IP, error) {
+	dsURL, err := url.Parse(serverURL)
+	if err != nil {
+		log.Errorf("findDSmDNS: error %v", err)
+		return "", "", nil, err
+	}
+	hostname := dsURL.Hostname()
+	var newURL string
+	var foundIP []net.IP
+
+	urls := strings.Split(serverURL, hostname)
+	if len(urls) != 2 {
+		return "", "", nil, fmt.Errorf("findDSmDNS: urls format error %v", urls)
+	}
+
+	// try to loop through the well-known services, workstation, http and https
+	// until success, for the local data store zeroconf
+	services := []string{
+		"workstation", "http", "https",
+	}
+
+	sub := dctx.subNetworkInstanceStatus
+	niItems := sub.GetAll()
+	ifs := findLocalBridges(niItems)
+	if len(ifs) == 0 {
+		return "", "", nil, fmt.Errorf("findDSmDNS: has no bridge interface")
+	}
+
+	for _, service := range services {
+		foundIP, err = queryService(ifs, hostname, service)
+		if err != nil {
+			return "", "", nil, err
+		}
+		if len(foundIP) > 0 {
+			break
+		}
+	}
+
+	if len(foundIP) == 0 {
+		return "", "", nil, fmt.Errorf("findDSmDNS: mDNS host not found")
```

Comment:
How about including the hostname in the error string?

Reply:
ok